builtin/providers/dockercloud/service.go

```diff
@@ -0,0 +1,313 @@
+					},
+					"rewritable": &schema.Schema{
+						Type:     schema.TypeBool,
+						Default:  true,
+						Optional: true,
+					},
+					"volumes_from": &schema.Schema{
+						Type:          schema.TypeString,
+						Optional:      true,
+						ConflictsWith: []string{"bindings.container_path"},
+					},
+				},
+			},
+		},
+		"env": &schema.Schema{
+			Type:     schema.TypeSet,
+			Optional: true,
+			Set:      envHash,
+			Elem: &schema.Resource{
+				Schema: map[string]*schema.Schema{
+					"key": &schema.Schema{
+						Type:     schema.TypeString,
+						Required: true,
+					},
+					"value": &schema.Schema{
+						Type:     schema.TypeString,
+						Required: true,
+					},
+				},
+			},
+		},
+		"links": &schema.Schema{
+			Type:     schema.TypeSet,
+			Optional: true,
+			Set:      linksHash,
+			Elem: &schema.Resource{
+				Schema: map[string]*schema.Schema{
+					"to": &schema.Schema{
+						Type:     schema.TypeString,
+						Required: true,
+					},
+					"name": &schema.Schema{
+						Type:     schema.TypeString,
+						Optional: true,
+					},
+				},
+			},
+		},
+		"ports": &schema.Schema{
+			Type:     schema.TypeSet,
+			Optional: true,
+			Set:      portsHash,
+			Elem: &schema.Resource{
+				Schema: map[string]*schema.Schema{
+					"internal": &schema.Schema{
+						Type:     schema.TypeInt,
+						Required: true,
+					},
+					"external": &schema.Schema{
+						Type:     schema.TypeInt,
+						Optional: true,
+					},
+					"protocol": &schema.Schema{
+						Type:     schema.TypeString,
+						Default:  "tcp",
+						Optional: true,
+					},
+				},
+			},
+		},
+		"redeploy_on_change": &schema.Schema{
+			Type:     schema.TypeBool,
+			Optional: true,
+		},
+		"reuse_existing_volumes": &schema.Schema{
+			Type:     schema.TypeBool,
+			Default:  true,
+			Optional: true,
+		},
+		"uri": &schema.Schema{
+			Type:     schema.TypeString,
+			Computed: true,
+		},
+		"public_dns": &schema.Schema{
+			Type:     schema.TypeString,
+			Computed: true,
+		},
+	}
+}
+
+func newServiceCreateRequest(d *schema.ResourceData) dockercloud.ServiceCreateRequest {
+	opts := dockercloud.ServiceCreateRequest{
+		Name:  d.Get("name").(string),
+		Image: d.Get("image").(string),
+	}
+
+	if attr, ok := d.GetOk("entrypoint"); ok {
+		opts.Entrypoint = attr.(string)
+	}
+
+	if attr, ok := d.GetOk("command"); ok {
+		opts.Run_command = attr.(string)
+	}
+
+	if attr, ok := d.GetOk("workdir"); ok {
+		opts.Working_dir = attr.(string)
+	}
+
+	if attr, ok := d.GetOk("pid"); ok {
+		opts.Pid = attr.(string)
+	}
+
+	if attr, ok := d.GetOk("net"); ok {
+		opts.Net = attr.(string)
+	}
+
+	if attr, ok := d.GetOk("deployment_strategy"); ok {
+		opts.Deployment_strategy = attr.(string)
+	}
+
+	if attr, ok := d.GetOk("autorestart"); ok {
+		opts.Autorestart = attr.(string)
+	}
+
+	if attr, ok := d.GetOk("autodestroy"); ok {
+		opts.Autodestroy = attr.(string)
+	}
+
+	if attr, ok := d.GetOk("autoredeploy"); ok {
```

Comment:
So bools and d.GetOk are not the best way IMO - if this value is false, it will never evaluate out of the d.GetOk func. We usually suggest to add a Default to the schema and then use d.Get instead